### ISSUE TITLE
[java] False positive on AvoidCatchingGenericException for try-with-resource blocks

### ISSUE THREAD

**Issue by mryan43** [pmd/pmd#2135]
**Affects PMD Version:** 6.13

**Rule:** [AvoidCatchingGenericException](https://pmd.github.io/latest/pmd_rules_java_design.html#avoidcatchinggenericexception)

**Description:**

`AutoClosable` defines by default `void close() throws Exception`. When implementing it, the developer may remove the throws / clamp down the exception type, but if he doesn't, all `try-with-resources` will force to catch `Exception`.

PMD should check if the exception type being catched is actually required or can be avoided before reporting a violation.

**Code Sample demonstrating the issue:**

```java
private class MyResource implements AutoCloseable {
  public MyResource() throws MyException {
  }

  public void close() throws Exception {
    ...
  }
}

// somewhere in the code
try(MyResource myResource = new MyResource()){
  ...
} catch (Exception e) {
  // Handle Exception
}
```

**Running PMD through:** *Maven*

**Comment by jsotuyod** [comment]
@mryan43 thanks for the report.

There is no limitation on what you can catch when using try-with-resources, even if the exception is throws by the constructor.

However, you are most probably being forced to catch `Exception` since `AutoClosable.close()` is by default defined as `throws Exception`, so unless you clamp down the exception type when implementing the method, that's what you are going to catch. I'm updating the issue to reflect this.

The rule however, has currently no knowledge of what exception types are actually being thrown… we *may*  whitelist `Exception` with try-with-resources, but that would introduce false negatives. An alternative approach would be to reimplement the rule in Java and manually check throw types for each call using reflection.

**Comment by mryan43** [comment]
As described in the code snippet (before you edited it), I am clamping down the exception type in the close method. The java compiler still requires me to catch Exception. Can you please give me an example code where close clamps the exception type, and the java compiler doesn't complain, and that doesn't trigger AvoidCatchingGenericException ?

**Comment by jsotuyod** [comment]
@mryan43 sorry for the late response…

`thows` is not part of a method signature in Java. This means, you can write stuff such as:

```java
public class NonThrowingClosable implements AutoClosable {
  @Override
  public void close() {
    // yes, this is an override, but we don't throw Exception, as this implementation is a nop
    // this is exactly what Java's BaseStream does for instance
  }
}
```

```java
public class IOExceptionThrowingClosable implements AutoClosable {
  @Override
  public void close() throws IOException {
    // yes, this is an override, but we throw a more specific exception type
    // this is exactly what Java's InputStream does for instance
  }
}
```

which in turn allows you to do:

```java
try (NonThrowingClosable res = new NonThrowingClosable()) {
  // do stuff
}
// no catch needed!
```

```java
try (IOExceptionThrowingClosable res = new IOExceptionThrowingClosable()) {
  // do stuff
} catch (IOException e) {
  // we catch a more specific exception type
}
```

It's really up to you as the implementor of `AutoClosable` to decide which (if any) exceptions to throw upon close.

**Comment by mryan43** [comment]
you're right.

I tested it again and I wasn't able to reproduce my original issue.

Closing this, sorry for the inconvenience
